### PR TITLE
Add support for custom Technic client IDs

### LINKED ISSUE #2775: Add Technic Solder Client ID to APIs tab
### Role

Testing private/unlisted modpacks on Technic

### Suggestion

Allow setting Technic Client ID

### Benefit

So I and others can download/test private and unlisted Technic packs

### This suggestion is unique

- [X] I have searched the issue tracker and did not find an issue describing my suggestion, especially not one that has been rejected.

### You may use the editor below to elaborate further.

I believe this is a feature of Technic Solder

Launcher v3 is GPL-3.0 so any reimplementation shouldn't have license problems

Relevant lines:

https://github.com/TechnicPack/LauncherV3/blob/3c4dd24983f2d12b73cb6c19b8fe1707e2040e13/src/main/java/net/technicpack/solder/http/HttpSolderApi.java#L68

https://github.com/TechnicPack/LauncherV3/blob/3c4dd24983f2d12b73cb6c19b8fe1707e2040e13/src/main/java/net/technicpack/solder/http/HttpSolderPackApi.java#L67

Format is a pretty standard UUID

Screenshot of the setting in the launcher (with my id censored):
image

Thinking about implementing this myself if noone else is interested

### PR DESCRIPTION
**Pull request by Trial97**
fixes #2775

### CODE REVIEW

File: launcher/Application.cpp [hunk 1]
```diff
@@ -777,6 +777,9 @@ Application::Application(int& argc, char** argv) : QApplication(argc, argv)
         // FTBApp instances
         m_settings->registerSetting("FTBAppInstancesPath", "");
 
+        // Custom Techinc Client ID
+        m_settings->registerSetting("TechincClientID", "");
```
Comment (getchoo) [inline launcher/Application.cpp]:
```suggestion
        // Custom Technic Client ID
        m_settings->registerSetting("TechnicClientID", "");
```

File: launcher/ui/pages/global/APIPage.cpp [hunk 2]
```diff
@@ -143,6 +143,7 @@ void APIPage::loadSettings()
     ui->modrinthToken->setText(modrinthToken);
     QString customUserAgent = s->get("UserAgentOverride").toString();
     ui->userAgentLineEdit->setText(customUserAgent);
+    ui->techicClientID->setText(s->get("TechincClientID").toString());
```
Comment (getchoo) [inline launcher/ui/pages/global/APIPage.cpp]:
```suggestion
    ui->techicClientID->setText(s->get("TechnicClientID").toString());
```

File: launcher/ui/pages/global/APIPage.cpp [hunk 3]
```diff
@@ -172,6 +173,7 @@ void APIPage::applySettings()
     QString modrinthToken = ui->modrinthToken->text();
     s->set("ModrinthToken", modrinthToken);
     s->set("UserAgentOverride", ui->userAgentLineEdit->text());
+    s->set("TechincClientID", ui->techicClientID->text());
```
Comment (getchoo) [inline launcher/ui/pages/global/APIPage.cpp]:
```suggestion
    s->set("TechincClientID", ui->technicClientID->text());
```

File: launcher/ui/pages/global/APIPage.ui [hunk 4]
```diff
@@ -288,6 +288,36 @@
          </layout>
         </widget>
        </item>
+       <item>
+        <widget class="QGroupBox" name="groupBox">
+         <property name="title">
+          <string>Techinc Client ID</string>
```
Comment (getchoo) [inline launcher/ui/pages/global/APIPage.ui]:
```suggestion
          <string>Technic Client ID</string>
```

File: launcher/ui/pages/global/APIPage.ui [hunk 5]
```diff
@@ -288,6 +288,36 @@
...
+           <widget class="QLineEdit" name="techicClientID">
+            <property name="placeholderText">
+             <string>(None)</string>
+            </property>
+           </widget>
+          </item>
+          <item>
+           <widget class="QLabel" name="label_12">
+            <property name="text">
+             <string>Enter a custom GUID client ID for Techinc here.</string>
```
Comment (getchoo) [inline launcher/ui/pages/global/APIPage.ui]:
```suggestion
             <string>Enter a custom GUID client ID for Technic here.</string>
```

File: launcher/ui/pages/modplatform/technic/TechnicModel.cpp [hunk 6]
```diff
@@ -154,6 +154,10 @@ void Technic::ListModel::performSearch()
             QString("%1search?build=%2&q=%3").arg(BuildConfig.TECHNIC_API_BASE_URL, BuildConfig.TECHNIC_API_BUILD, currentSearchTerm);
         searchMode = List;
     }
+    auto clientId = APPLICATION->settings()->get("TechincClientID").toString();
```
Comment (getchoo) [inline launcher/ui/pages/modplatform/technic/TechnicModel.cpp]:
```suggestion
    auto clientId = APPLICATION->settings()->get("TechnicClientID").toString();
```

**Review by getchoo — CHANGES_REQUESTED** [review]
s/techinc/technic/g

File: launcher/ui/pages/global/APIPage.ui [hunk 7]
```diff
@@ -288,6 +288,36 @@
...
+         <layout class="QVBoxLayout" name="verticalLayout_9">
+          <item>
+           <widget class="QLabel" name="label_11">
+            <property name="text">
+             <string>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;Note: you only need to set this to access private data.&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</string>
+            </property>
+           </widget>
+          </item>
+          <item>
+           <widget class="QLineEdit" name="techicClientID">
```
Comment (Trial97) [inline launcher/ui/pages/global/APIPage.ui]:
```suggestion
           <widget class="QLineEdit" name="technicClientID">
```

File: launcher/ui/pages/global/APIPage.cpp [hunk 8]
```diff
@@ -172,6 +173,7 @@ void APIPage::applySettings()
     QString modrinthToken = ui->modrinthToken->text();
     s->set("ModrinthToken", modrinthToken);
     s->set("UserAgentOverride", ui->userAgentLineEdit->text());
+    s->set("TechincClientID", ui->technicClientID->text());
```
Comment (TheKodeToad) [inline launcher/ui/pages/global/APIPage.cpp]:
```suggestion
    s->set("TechnicClientID", ui->technicClientID->text());
```

**Review by TheKodeToad — APPROVED** [review]

**Review by getchoo — APPROVED** [review]

**Pull request merged**